File: packages/moderation/src/validation-rules/rules/disclosures.ts
import type { Labrinth } from '@modrinth/api-client'
import { defineMessages } from '@modrinth/ui/i18n'
import { formatProjectTypeSentence } from '@modrinth/ui/src/utils/common-messages.ts'

import type { Nag, ProjectValidationContext } from '../../types/nags.ts'
import { evaluateRules } from '../evaluate-rules.ts'
import { hasProjectTextHtmlFormatting } from '../text.ts'
import { toFieldMessages } from '../to-field-messages.ts'
import { toNags } from '../to-nags.ts'
import type { FieldValidationMessage, ValidationRuleSet } from '../types.ts'

const messages = defineMessages({
	title: {
		id: 'nags.check-disclosures.title',
		defaultMessage: 'Check content disclosures',
	},
	description: {
		id: 'nags.check-disclosures.description',
		defaultMessage:
			'Make sure users are aware of any important details by filling in content disclosures that apply to your {type}.',
	},
	removeFormatting: {
		id: 'nags.disclosures-special-formatting.title',
		defaultMessage: 'Remove HTML from content disclosures',
	},
	specialFormatting: {
		id: 'nags.disclosures-special-formatting.description',
		defaultMessage: 'Content disclosures cannot contain HTML, as they display in plain text.',
	},
})

function getDisclosureText(disclosure: Labrinth.Projects.v3.ProjectDisclosure): string[] {
	switch (disclosure.type) {
		case 'ai_content':
		case 'advertisements':
		case 'epilepsy_triggers':
		case 'archived':
			return disclosure.note ? [disclosure.note] : []
		case 'system_interactions':
			return disclosure.note ? [disclosure.note] : []
		case 'telemetry':
			return disclosure.data_collected
		case 'derivative_work':
			return disclosure.sources.flatMap((source) =>
				source.note ? [source.label, source.note] : [source.label],
			)
		case 'paid_features':
			return disclosure.features
	}
}

export const projectDisclosureTextValidationRules = {
	'disclosures-special-formatting': {
		severity: 'error',
		evaluate: (disclosures) => ({
			valid: !disclosures.some((disclosure) =>
				getDisclosureText(disclosure).some(hasProjectTextHtmlFormatting),
			),
		}),
		presentation: {
			message: messages.specialFormatting,
			nag: { title: messages.removeFormatting, destination: 'disclosures' },
		},
	},
} satisfies ValidationRuleSet<readonly Labrinth.Projects.v3.ProjectDisclosure[]>

export function validateProjectDisclosures(
	disclosures: readonly Labrinth.Projects.v3.ProjectDisclosure[],
): FieldValidationMessage[] {
	return toFieldMessages(evaluateRules(disclosures, projectDisclosureTextValidationRules))
}

export const projectDisclosureValidationRules = {
	'check-disclosures': {
		severity: 'suggestion',
		evaluate: (context) => ({
			valid: false,
			values: { projectType: context.project.project_type },
		}),
		presentation: {
			message: messages.description,
			nag: {
				title: messages.title,
				destination: 'disclosures',
				formatValues: (values, formatMessage) => ({
					type: formatProjectTypeSentence(formatMessage, String(values.projectType)),
				}),
			},
		},
	},
} satisfies ValidationRuleSet<ProjectValidationContext>

export function getDisclosureNags(context: ProjectValidationContext): Nag[] {
	return toNags(evaluateRules(context, projectDisclosureValidationRules))
}

File: apps/labrinth/src/validate/project/disclosures.rs
use super::text::has_paired_html_formatting;
use super::{ProjectNag, ProjectNagKind, ProjectNagSeverity};
use crate::models::disclosures::ProjectDisclosure;
use crate::models::v2::projects::LegacyProject;

pub(super) fn validate(
    project: &crate::models::projects::Project,
    disclosures: Option<&[ProjectDisclosure]>,
) -> Vec<ProjectNag> {
    let (project_type, _) =
        LegacyProject::get_project_type(&project.project_types);
    let mut nags = vec![
        ProjectNag::new(
            ProjectNagKind::CheckDisclosures,
            ProjectNagSeverity::Suggestion,
        )
        .with_details(serde_json::json!({ "project_type": project_type })),
    ];

    if disclosures.is_some_and(|disclosures| {
        disclosures.iter().any(disclosure_has_paired_html)
    }) {
        nags.push(ProjectNag::new(
            ProjectNagKind::DisclosuresSpecialFormatting,
            ProjectNagSeverity::Required,
        ));
    }

    nags
}

fn disclosure_has_paired_html(disclosure: &ProjectDisclosure) -> bool {
    match disclosure {
        ProjectDisclosure::AiContent { note, .. }
        | ProjectDisclosure::Advertisements { note }
        | ProjectDisclosure::EpilepsyTriggers { note }
        | ProjectDisclosure::SystemInteractions { note, .. }
        | ProjectDisclosure::Archived { note } => {
            note.as_deref().is_some_and(has_paired_html_formatting)
        }
        ProjectDisclosure::Telemetry { data_collected, .. } => data_collected
            .iter()
            .any(|text| has_paired_html_formatting(text)),
        ProjectDisclosure::DerivativeWork { sources } => {
            sources.iter().any(|source| {
                has_paired_html_formatting(&source.label)
                    || source
                        .note
                        .as_deref()
                        .is_some_and(has_paired_html_formatting)
            })
        }
        ProjectDisclosure::PaidFeatures { features } => {
            features.iter().any(|text| has_paired_html_formatting(text))
        }
    }
}
